Add Voronoi options to include grain direction and anisotropy, add wood yard scene demo

Glavin001 opened the pull request:
| Before | After 1 | After 2 |
| --- | --- | --- |
| <img width="989" height="685" alt="image" src="https://github.com/user-attachments/assets/897d5153-f884-4f62-b010-ee3138003d6e" /> | <img width="993" height="686" alt="image" src="https://github.com/user-attachments/assets/e864cbc7-955f-49b7-8806-231b957eec0a" /> |  <img width="994" height="687" alt="image" src="https://github.com/user-attachments/assets/3d1e1318-b190-41d7-ba42-8b18ca7bf543" /> |


- Updated Voronoi options to include grain direction and anisotropy for improved material behavior during fractures.
- Enhanced DestructibleMesh to support anisotropic fracturing based on grain direction and anisotropy factor.
- Implemented a MaterialFactory method to load and cache wood materials for realistic rendering.
- Created a new WoodYardScene to showcase wood objects with anisotropic Voronoi fracturing, allowing for realistic splintering effects.
- Introduced wood textures from Polyhaven, including diffuse, normal, and ambient occlusion maps.

Glavin001 committed:
Add Voronoi options to include grain direction and anisotropy, add wood yard scene demo

- Introduced wood textures from Polyhaven, including diffuse, normal, and ambient occlusion maps.
- Implemented a MaterialFactory method to load and cache wood materials for realistic rendering.
- Created a new WoodYardScene to showcase wood objects with anisotropic Voronoi fracturing, allowing for realistic splintering effects.
- Enhanced DestructibleMesh to support anisotropic fracturing based on grain direction and anisotropy factor.
- Updated Voronoi options to include grain direction and anisotropy for improved material behavior during fractures.

dgreenheck commented on lib/src/fracture/VoronoiCell.ts:
```diff
@@ -42,6 +42,58 @@ export function computeBisectingPlane(
   return { origin, normal };
 }
 
+/**
```
Can you generate unit tests for this file please?

Glavin001 replied:
@dgreenheck Done.

dgreenheck commented on lib/src/fracture/VoronoiFracture.ts:
```diff
@@ -80,6 +80,10 @@ function voronoiFracture3D(
     );
   }
 
+  // Extract anisotropic parameters if provided
```
And this file as well. Sorry, should have already added them!

Glavin001 replied:
@dgreenheck Done.

dgreenheck commented on lib/src/fracture/VoronoiCell.ts:
```diff
@@ -42,6 +42,58 @@ export function computeBisectingPlane(
+  // Anisotropic factor: (1 - 1/A²)
+  // When A = 1, factor = 0, giving standard isotropic Voronoi
+  // When A > 1, factor approaches 1, giving maximum elongation
+  const factor = 1 - 1 / (anisotropy * anisotropy);
```
Need a divide by zero check here. Does an anisotropy of < 1 make sense? If not, may be good to add checks in `VoronoiFractureOptions` as well.

Glavin001 replied:
@dgreenheck Done, made it clamp to `>= 1.0`.

Glavin001 committed:
Enhance VoronoiFractureOptions and related tests for anisotropy handling

- Updated VoronoiFractureOptions to clamp anisotropy values to a minimum of 1.0, ensuring physical validity.
- Added comprehensive unit tests for anisotropy behavior, including clamping and normalization of grain direction.
- Implemented tests for anisotropic bisecting plane calculations, validating expected behavior across various anisotropy values.
- Introduced new test cases for VoronoiCell and VoronoiFracture to ensure robustness in anisotropic scenarios.

Glavin001 commented on demo/src/app.ts:
```diff
@@ -13,6 +13,8 @@ import { SlicingScene } from "./scenes/SlicingScene";
 import { BrickWallScene } from "./scenes/BrickWallScene";
 import { RefractureScene } from "./scenes/RefractureScene";
 import { KintsugiScene } from "./scenes/KintsugiScene";
+// New material demo scenes
```
```suggestion
```

Glavin001 committed:
Apply suggestion from @Glavin001

Glavin001 commented:
Anything else I can do for this PR, @dgreenheck ? 😃 

Love this library and your posts on Twitter/X, btw! Thank you!
